go-controller/pkg/clustermanager/userdefinednetwork/controller_test.go

```diff
@@ -17,535 +19,854 @@ import (
+					By("create test new namespaces with new selection label")
+					newNsLabelValue := "us"
+					newNsLabel := map[string]string{testLabelKey: newNsLabelValue}
+					newNsNames := []string{"black", "gray"}
+					for _, nsName := range newNsNames {
+						ns := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: nsName, Labels: newNsLabel}}
+						ns, err := cs.KubeClient.CoreV1().Namespaces().Create(context.Background(), ns, metav1.CreateOptions{})
+						Expect(err).NotTo(HaveOccurred())
+					}
 
-		It("when UDN is being deleted, NAD exist, 2 pods using UDN, should remove finalizers once no pod uses the network", func() {
-			udn := testsUDNWithDeletionTimestamp(time.Now())
-			udn, err := udnClient.K8sV1().UserDefinedNetworks(udn.Namespace).Create(context.Background(), udn, metav1.CreateOptions{})
-			Expect(err).NotTo(HaveOccurred())
+					By("add new label to CR namespace-selector")
+					p := fmt.Sprintf(`[{"op": "add", "path": "./spec/namespaceSelector/matchExpressions/0/values/-", "value": "%s"}]`, newNsLabelValue)
+					cudn, err := cs.UserDefinedNetworkClient.K8sV1().ClusterUserDefinedNetworks().Patch(context.Background(), cudnName, types.JSONPatchType, []byte(p), metav1.PatchOptions{})
+					Expect(err).NotTo(HaveOccurred())
+					Expect(cudn.Spec.NamespaceSelector.MatchExpressions).To(Equal([]metav1.LabelSelectorRequirement{{
+						Key:      testLabelKey,
+						Operator: metav1.LabelSelectorOpIn,
+						Values:   []string{testLabelValue, newNsLabelValue},
+					}}))
+
+					Eventually(func() []metav1.Condition {
+						var err error
+						cudn, err = cs.UserDefinedNetworkClient.K8sV1().ClusterUserDefinedNetworks().Get(context.Background(), cudnName, metav1.GetOptions{})
+						Expect(err).NotTo(HaveOccurred())
+						return normalizeConditions(cudn.Status.Conditions)
+					}).Should(Equal([]metav1.Condition{{
+						Type:    "NetworkReady",
+						Status:  "True",
+						Reason:  "NetworkAttachmentDefinitionReady",
+						Message: "NetworkAttachmentDefinition has been created in following namespaces: [black, gray, green, yellow]",
+					}}), "status should report NAD exist in existing and new labeled namespaces")
+					for _, nsName := range append(connectedNsNames, newNsNames...) {
+						nads, err := cs.NetworkAttchDefClient.K8sCniCncfIoV1().NetworkAttachmentDefinitions(nsName).List(context.Background(), metav1.ListOptions{})
+						Expect(err).NotTo(HaveOccurred())
+						Expect(nads.Items).To(Equal([]netv1.NetworkAttachmentDefinition{*testClusterUdnNAD(cudnName, nsName)}),
+							"NAD should exist in existing and new labeled namespaces")
+					}
+				})
+
+				It("when CR selector has selection removed, should delete stale NADs in previously matching namespaces", func() {
+					By("remove test label value from namespace-selector")
+					p := fmt.Sprintf(`[{"op": "replace", "path": "./spec/namespaceSelector/matchExpressions/0/values", "value": [""]}]`)
```

Comment:
this one is a bit hard to follow, can you just use golang struct and call Update instead of patch?

Reply:
Done

Reply:
I found 8 more patch calls, do you think we can replace all of them with updates?

Reply:
Done

I left usable of Patch in places the subject object is not available and require fetching.